azurerm/resource_arm_iothub_resource_provider.go

```diff
@@ -0,0 +1,264 @@
+
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	rg := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	updateIoT := ""
+
+	res, err := iothubClient.CheckNameAvailability(ctx, devices.OperationInputs{
+		Name: &name,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !*res.NameAvailable {
+		getETag, err := iothubClient.Get(ctx, rg, name)
+		if err != nil {
+			return errors.New(string(res.Reason))
+		}
+		updateIoT = *getETag.Etag
+	}
+
+	location := d.Get("location").(string)
+
+	subscriptionID := armClient.subscriptionId
+	skuInfo := expandAzureRmIotHubSku(d)
+
+	desc := devices.IotHubDescription{
+		Resourcegroup:  &rg,
+		Name:           &name,
+		Location:       &location,
+		Subscriptionid: &subscriptionID,
+		Sku:            &skuInfo,
+	}
+
+	if tagsI, ok := d.GetOk("tags"); ok {
+		tags := tagsI.(map[string]interface{})
+		desc.Tags = *expandTags(tags)
+	}
+
+	future, err := iothubClient.CreateOrUpdate(ctx, rg, name, desc, updateIoT)
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, iothubClient.Client)
+	if err != nil {
+		return fmt.Errorf("Error creating or updating IotHub %q (Resource Group %q): %+v", name, rg, err)
+	}
+
+	desc, err = iothubClient.Get(ctx, rg, name)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*desc.ID)
+	return resourceArmIotHubRead(d, meta)
+
+}
+
+func expandAzureRmIotHubSku(d *schema.ResourceData) devices.IotHubSkuInfo {
+	skuList := d.Get("sku").([]interface{})
+	skuMap := skuList[0].(map[string]interface{})
+	cap := int64(skuMap["capacity"].(int))
+
+	name := skuMap["name"].(string)
+	tier := skuMap["tier"].(string)
+
+	return devices.IotHubSkuInfo{
+		Name:     devices.IotHubSku(name),
+		Tier:     devices.IotHubSkuTier(tier),
+		Capacity: &cap,
+	}
+
+}
+
+func resourceArmIotHubRead(d *schema.ResourceData, meta interface{}) error {
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	iothubName := id.Path["IotHubs"]
+	desc, err := iothubClient.Get(ctx, id.ResourceGroup, iothubName)
+
+	if err != nil {
+		return err
+	}
+
+	properties := desc.Properties
+
+	keysResp, err := iothubClient.ListKeys(ctx, id.ResourceGroup, iothubName)
+	keyList := keysResp.Response()
+
+	var keys []map[string]interface{}
+	for _, key := range *keyList.Value {
+		keyMap := make(map[string]interface{})
+		keyMap["key_name"] = *key.KeyName
+		keyMap["primary_key"] = *key.PrimaryKey
+		keyMap["secondary_key"] = *key.SecondaryKey
+		keyMap["permissions"] = string(key.Rights)
+		keys = append(keys, keyMap)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error flattening `shared_access_policy` in IoTHub %q: %+v", iothubName, err)
+	}
+
+	d.Set("shared_access_policy", keys)
+	d.Set("hostname", properties.HostName)
+	d.Set("type", desc.Type)
+	flattenAndSetTags(d, &desc.Tags)
```

Comment:
can we also set the `sku` block? This should be returned as part of the properties object